### PR TITLE
Change edge types to numbers

### PR DESCRIPTION
**Pull request by thebriando**
# ↪️ Pull Request

Eventually we'll work on moving the edge list in the Graph to a buffer, but we need to change edge types to numbers first in order to do that.

## 🚨 Test instructions

* Updated unit tests to reflect the new edge type format
* Tested on a large project

## ✔️ PR Todo

- [x] Added/updated unit tests for this change
- [x] Filled out test instructions (In case there aren't any unit tests)
- [ ] Included links to related issues/PRs

### CODE REVIEW

**Comment by parcel-benchmark** [comment]
## Benchmark Results
<details><summary>Kitchen Sink ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 2.03s | +17.00ms |
| Cached | 307.00ms | -6.00ms |

#### Cold Bundles

*No bundle changes detected.*

#### Cached Bundles

*No bundle changes detected.*

</p></details>

<details><summary>React HackerNews ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 10.40s | -246.00ms |
| Cached | 503.00ms | +1.00ms |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/logo.c5bb83f1.png | 246.00b | +0.00b | 5.65s | +627.00ms ⚠️ |

#### Cached Bundles

*No bundle changes detected.*

</p></details>

<details><summary>AtlasKit Editor ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 1.14m | -543.00ms |
| Cached | 1.78s | -67.00ms |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/index.774149b9.js | 1.78mb | -2.00b 🚀 | 19.65s | -389.00ms |
| dist/index.3f9d78fb.js | 694.72kb | +15.00b ⚠️ | 52.53s | -852.00ms |
| dist/EmojiPickerComponent.0491fb2b.js | 146.67kb | +0.00b | 31.96s | +10.24s ⚠️ |
| dist/esm.522ea207.js | 33.15kb | +0.00b | 31.96s | +10.24s ⚠️ |
| dist/DatePicker.9038cf9f.js | 22.98kb | +0.00b | 31.96s | +10.24s ⚠️ |
| dist/js.0f0bb621.js | 17.25kb | +0.00b | 51.93s | +12.49s ⚠️ |
| dist/png-chunks-extract.94b5b9fc.js | 3.58kb | +0.00b | 51.93s | +12.49s ⚠️ |
| dist/feedback.e18b45f0.js | 1.77kb | +0.00b | 31.95s | +10.24s ⚠️ |
| dist/workerHasher.567cfc57.js | 1.63kb | +0.00b | 31.96s | +10.24s ⚠️ |
| dist/heading6.b3b946d1.js | 1.51kb | +0.00b | 31.95s | +10.24s ⚠️ |
| dist/heading5.e617db66.js | 1.38kb | +0.00b | 31.95s | +10.24s ⚠️ |
| dist/expand.1f17de7f.js | 1.29kb | +0.00b | 31.95s | +10.24s ⚠️ |
| dist/heading4.62b9719b.js | 1.27kb | +0.00b | 31.95s | +10.24s ⚠️ |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/index.e288bc82.js | 1.78mb | +2.00b ⚠️ | 19.75s | -67.00ms |
| dist/index.96a60c6b.js | 694.71kb | -15.00b 🚀 | 52.75s | -289.00ms |
| dist/Modal.2b19bf46.js | 45.31kb | +15.00b ⚠️ | 52.17s | -261.00ms |

</p></details>

<details><summary>Three.js ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 7.42s | -61.00ms |
| Cached | 444.00ms | +30.00ms ⚠️ |

#### Cold Bundles

*No bundle changes detected.*

#### Cached Bundles

*No bundle changes detected.*

</p></details>

**[Click here to view a detailed benchmark overview.](https://benchmark.parceljs.org/benchmark/309776906521150018)**

**Comment by joeyslater** [comment]
You may want to look at `dumpGraphToGraphViz` as I believe the colors won't map anymore

File: packages/core/core/src/RequestTracker.js [hunk 1]
```diff
@@ -37,6 +37,16 @@ import {
   ERROR,
 } from './constants';
 
+export const RequestGraphEdgeTypes = {
+  subrequest: 1,
+  invalidated_by_update: 2,
```
Comment (wbinnssmith) [inline packages/core/core/src/RequestTracker.js]: nit: camelCase?

File: packages/core/core/src/Graph.js [hunk 2]
```diff
@@ -339,7 +344,7 @@ export default class Graph<TNode: Node, TEdgeType: string | null = null> {
   traverse<TContext>(
     visit: GraphVisitor<TNode, TContext>,
     startNode: ?TNode,
-    type: TEdgeType | null | Array<TEdgeType | null> = null,
+    type: TEdgeType | NullEdgeType | Array<TEdgeType> = 0,
```
Comment (wbinnssmith) [inline packages/core/core/src/Graph.js]:
```suggestion
    type: TEdgeType | NullEdgeType | Array<TEdgeType | NullEdgeType> = 0,
```

File: packages/core/core/src/Graph.js [hunk 3]
```diff
@@ -6,18 +6,19 @@ import type {TraversalActions, GraphVisitor} from '@parcel/types';
 import assert from 'assert';
 import nullthrows from 'nullthrows';
 
-export type GraphOpts<TNode, TEdgeType: string | null = null> = {|
+type NullEdgeType = 0;
```
Comment (wbinnssmith) [inline packages/core/core/src/Graph.js]: I wonder if we should not support the null edge type at all, and just have users/extenders always specify the types of edges the way they do nodes.

Comment (wbinnssmith) [reply packages/core/core/src/Graph.js]: cc @devongovett @mischnic

**Comment by thebriando** [comment]
Closing this for now since this will be involved in a different task for buffer-backed edges in the graph.

**Comment by wbinnssmith** [comment]
Why did we close this? Couldn't this be landed separately from upcoming work?

**Comment by thebriando** [comment]
> Why did we close this? Couldn't this be landed separately from upcoming work?

I _think_ I closed this because we thought we were close to opening a PR for the main work 😞  which includes the numeric edge types, and landing this by itself didn't have any significant performance impact.

**Comment by wbinnssmith** [comment]
> didn't have any significant performance impact.

It will make the work reviewable though, yeah?

**Comment by thebriando** [comment]
> It will make the work reviewable though, yeah?

Yeah good point, I reopened it.

**Comment by wbinnssmith** [comment]
This should be able to land before #6922 I believe. Looks like there's a small lint issue.

**Comment by thebriando** [comment]
> This should be able to land before #6922 I believe. Looks like there's a small lint issue.

Should be fixed now 😄

**Review by devongovett — APPROVED** [review]

**Pull request merged**